secret_store/src/listener/service_contract_listener.rs

```diff
@@ -0,0 +1,699 @@
+			.map_err(Into::into)
+	}
+
+	/// Restore server key.
+	fn restore_server_key(data: &Arc<ServiceContractListenerData>, server_key_id: &ServerKeyId) -> Result<Public, String> {
+		data.key_storage.get(server_key_id)
+			.map_err(|e| format!("{}", e))
+			.and_then(|ks| ks.ok_or("missing key".to_owned()))
+			.map(|ks| ks.public)
+	}
+
+	/// Publish server key.
+	fn publish_server_key(data: &Arc<ServiceContractListenerData>, server_key_id: &ServerKeyId, server_key: &Public) -> Result<(), String> {
+		data.contract.publish_server_key(server_key_id, server_key)
+	}
+}
+
+impl Drop for ServiceContractListener {
+	fn drop(&mut self) {
+		if let Some(service_handle) = self.service_handle.take() {
+			self.data.tasks_queue.shutdown();
+			// ignore error as we are already closing
+			let _ = service_handle.join();
+		}
+	}
+}
+
+impl ChainNotify for ServiceContractListener {
+	fn new_blocks(&self, _imported: Vec<H256>, _invalid: Vec<H256>, enacted: Vec<H256>, _retracted: Vec<H256>, _sealed: Vec<H256>, _proposed: Vec<Bytes>, _duration: u64) {
+		let enacted_len = enacted.len();
+		if enacted_len == 0 {
+			return;
+		}
+
+		self.data.contract.update();
+		if !self.data.contract.is_actual() {
+			return;
+		}
+
+		let reason = "enacted.len() != 0; qed";
+		self.process_service_contract_events(
+			enacted.first().expect(reason).clone(),
+			enacted.last().expect(reason).clone());
+
+		// schedule retry if received enough blocks since last retry
+		// it maybe inaccurate when switching syncing/synced states, but that's ok
+		if self.data.last_retry.fetch_add(enacted_len, Ordering::Relaxed) >= RETRY_INTEVAL_BLOCKS {
+			self.data.tasks_queue.push(::std::iter::once(ServiceTask::Retry));
+			self.data.last_retry.store(0, Ordering::Relaxed);
+		}
+	}
+}
+
+impl ClusterSessionsListener<GenerationSession> for ServiceContractListener {
+	fn on_session_removed(&self, session: Arc<GenerationSession>) {
+		// only publish when the session is started by another node
+		// when it is started by this node, it is published from process_service_task
+		if !is_processed_by_this_key_server(&*self.data.key_server_set, &*self.data.self_key_pair, &session.id()) {
+			// by this time sesion must already be completed - either successfully, or not
+			debug_assert!(session.is_finished());
+
+			// ignore result - the only thing that we can do is to log the error
+			let _ = session.wait(Some(Default::default()))
+				.map_err(|e| format!("{}", e))
+				.and_then(|server_key| Self::publish_server_key(&self.data, &session.id(), &server_key))
+				.map(|_| {
+					trace!(target: "secretstore", "{}: completed foreign GenerateServerKey({}) request",
+						self.data.self_key_pair.public(), session.id());
+					()
+				})
+				.map_err(|error| {
+					warn!(target: "secretstore", "{}: failed to process GenerateServerKey({}) request with: {}",
+						self.data.self_key_pair.public(), session.id(), error);
+					error
+				});
+		}
+	}
+}
+
+impl TasksQueue {
```

Comment:
That's a pretty elegant structure, consider putting it to a separate module and making it generic?

Reply:
Do you mean moving this to the local module of the SecretStore? Or somewhere in utils? I'm for the first scenario, if you confirm

Reply:
A local module is fine for now, it will clean up this one and it will be easier to re-use in the future.